### PR TITLE
[WIP][CR]Snow Overhaul

### PR DESCRIPTION
**Pull request by ghost**
#### Summary
```SUMMARY: Features "Adds a proper snow-fall mechanic"```

#### Purpose of change
Make snow an actual thing in-game beyond just a visual change of tiles on day 1 of Winter.

#### Describe the solution
1.Snow will now fall when the weather is snowing, the game will track how long it snows for and how heavy the snowfall is, and after a certain threshold deep snow piles will accumulate.
2. This snow will slowly fill up the map , it can also melt slowly if the temps go higher.
3. This snow, when settled will cause snow glare when its sunny.
4. This snow , when settled, will be slow to walk through, unless you have snowshoes ( a makeshift item is craftable ) 
5. Walking in the snow can make your feet cold - if you have no shoes, thats bad, if you have some shoes or mutated feet, its not as bad, if you have waterproof boots, its fine.
6. You can shovel the snow to remove it.
7. You can examine the snow to collect some snow, which can be used in the recipe to make water.
8. Sleeping on the snow would be very cold.
9. You can throw the snowball at zombies, if you want.
10. Someone else could make a recipe for a snowman construction if they wanted.

Problems I need help/advice/feedback on : 
1.

The visual aspect of it.
This is tricky, the graphical tiles for... deadpeople tileset for example, that have snow on, ( like t_grass_season_winter )
they do not correspond to any particular terrain.
Instead what happens before this PR, is that when it is winter that tile is automatically selected instead of t_grass.
I may need someone like @SomeDeadGuy or someone familiar with the tilesets , to separate out the various winter tiles, or to help me work out how best to use them to avoid graphical mismatches.
It seems to work passably well at the moment though.

Due to the tricky way seasonal graphical variations are done, I can either make it all ascii fallback, or make it work only on deadpeople tileset, I cant seem to find a solution that works for curses AND all tilesets.

2. 

Driving - not in this PR, but I will be looking at traction on snow, and adding snow chains etc.

3. 

the Snow item, this is tricky. currently, it dosnt melt. it just sits there, the game does not allow picking up frozen liquids.
I considered doing either a) making it water, then hacking together something that froze it instantly, and filled a container with water( frozen ) 
or b) making the snow ( frozen ) and then converting it to another item ( water ) when it "melted"
doing b ) would mean probably implementing a whole melting and evaporating code entirely.
doing a) would involve coding picking up frozen objects.

for the time being I went with c) and just made snow a weird item that never melts after you pick it up.
4. 

I would like to add trails/footprints in the snow that people/monsters/vehicle leave. not sure how to do that.

--------------------------------------------
I may need pointers and advice on how best to proceed with either of the two options. 

I need this to be tested thoroughly.

Ive done as much as I can, but it is so wide-ranging, I am sure I've forgotten something, I'd love people to think about all the things this may affect, mutations, bionics,  usages, potential bugs, problems, and list them to me.

#### Describe alternatives you've considered
Lots

#### Additional context
Little video of some snow falling whilst waiting ( time sped up )

### CODE REVIEW

**Comment by Night-Pryanik** [comment]
> You can throw the snowball at zombies

AT LAST!

**Comment by SomeDeadGuy** [comment]
Interesting!
Hit me up on Discord at SomeDeadGuy#3894 if you want help with graphical side of things.

File: src/cata_tiles.cpp [hunk 1]
```diff
@@ -1654,9 +1655,18 @@ const tile_type *cata_tiles::find_tile_with_season( std::string &id )
...
+    if( ( seasonnum == 3 ) && ( id == "t_grass" || id == "t_grass_dead" || id == "t_grass_long" ||
+                                id == "t_dirt" || id == "t_dirtmound" ) ) {
+        seasonnum = 2;
+    } else if( seasonnum == 3 && map::snowfall < 10000 ) {
+        seasonnum = 2;
+    }
+    if( seasonnum == 2 && map::snowfall > 10000 && !( id == "t_grass" || id == "t_grass_dead" ||
+            id == "t_grass_long" || id == "t_dirt" || id == "t_dirtmound" ) ) {
+        seasonnum = 3;
+    }
```
Comment (codemime) [inline src/cata_tiles.cpp]: Can the whole snippet be considered a temporary hack? The proper way of doing this is by means of the tileset jsons. Meanwhile, it deserves a thorough comment describing why it's here and under what circumstances it can be removed.

File: src/map.h [hunk 2]
```diff
@@ -1299,6 +1300,8 @@ class map
          * If false, monsters are not spawned in view of player character.
          */
         void spawn_monsters( bool ignore_sight );
+        static int snowfall;
+        static int percent_settled;
```
Comment (codemime) [inline src/map.h]: Why are these here as static members, and not in `w_point`? They don't seem to belong to `map`.

File: src/iuse.cpp [hunk 3]
```diff
@@ -2488,6 +2488,36 @@ int iuse::makemound( player *p, item *it, bool t, const tripoint & )
...
+        add_msg( m_info, _( "You clear your sinuses." ) );
+        return 0;
+    }
+    int moves;
+    if( g->m.furn( pnt ) == f_snow ) {
+        moves = MINUTES( 120 ) / it->get_quality( DIG ) * 100;
+    } else {
+        p->add_msg_if_player( _( "There isn't any snow there." ) );
+        return 0;
+    }
```
Comment (codemime) [inline src/iuse.cpp]: This kind of code is a bit dangerous. `moves` is mutable and can become uninitialized if someone decides to remove the `return` statement. 
```suggestion

    if( g->m.furn( pnt ) != f_snow ) {
        p->add_msg_if_player( _( "There isn't any snow there." ) );
        return 0;
    }
    const int moves = MINUTES( 120 ) / it->get_quality( DIG ) * 100;    
```

File: src/map.cpp [hunk 4]
```diff
@@ -5652,6 +5656,79 @@ void map::draw( const catacurses::window &w, const tripoint &center )
...
+    const w_point weatherPoint = *g->weather_precise;
+    if( g->weather == WEATHER_FLURRIES ) {
+        snowfall += 2;
+    } else if( g->weather == WEATHER_SNOW ) {
+        snowfall += 4;
+    } else if( g->weather == WEATHER_SNOWSTORM ) {
+        snowfall += 8;
+    } else {
+        snowfall -= std::max( 0, ( static_cast<int>( weatherPoint.temperature - 32 ) ) );
+    }
```
Comment (codemime) [inline src/map.cpp]: Can be converted to a `switch` statement inside a constant function that returns increments.
```c++
switch( g->weather ) {
case WEATHER_FLURRIES:
    return 2;
case WEATHER_SNOW:
    ...
}
```

File: src/map.cpp [hunk 5]
```diff
@@ -5652,6 +5656,79 @@ void map::draw( const catacurses::window &w, const tripoint &center )
...
+    snowfall = std::max( 0, snowfall );
+    snowfall = std::min( 172800, snowfall );
```
Comment (codemime) [inline src/map.cpp]: There's a `clamp` function for this.

File: src/map.cpp [hunk 6]
```diff
@@ -5652,6 +5656,79 @@ void map::draw( const catacurses::window &w, const tripoint &center )
...
+}
+
+void map::add_snow()
+{
+    if( g->weather >= WEATHER_FLURRIES && snowfall > 3000 ) {
```
Comment (codemime) [inline src/map.cpp]: Better be rewritten as a guard clause.
```c++
if( g->weather < WEATHER_FLURRIES || snowfall <= 3000 ) {
    return;
}
```

File: src/map.cpp [hunk 7]
```diff
@@ -5652,6 +5656,79 @@ void map::draw( const catacurses::window &w, const tripoint &center )
...
+        tripoint gp( 0, 0, 0 );
+        int total_grid = 0;
+        int snowfall_settled = 0;
+        int &gx = gp.x;
+        int &gy = gp.y;
```
Comment (codemime) [inline src/map.cpp]: Why this? `gp` isn't used anywhere else.

File: src/cata_tiles.cpp (hunk 1 above)
Comment (davidpwbrown) [reply src/cata_tiles.cpp]: This whole thing with the tilesets I'm gonna need to completely re-work, find a much better solution.
The winter versions of trees dont have visible fruit for example. 
And it needs to work for tilesets that don't have winter variants either.
I'll come back to this when I think of a better way.

File: src/map.cpp [hunk 8]
```diff
@@ -5652,6 +5656,79 @@ void map::draw( const catacurses::window &w, const tripoint &center )
...
+        int snowfactor = std::min( 15000, snowfall );
+        snowfactor = 15001 - snowfactor;
```
Comment (codemime) [inline src/map.cpp]: Can be rewritten as:
```c++
const int snowfactor = std::max( 15000 - snowfall, 1 );
```
Your implementation would yield `15001` when snowfall is `0`. Is that intentional?

File: src/map.cpp [hunk 9]
```diff
@@ -5652,6 +5656,79 @@ void map::draw( const catacurses::window &w, const tripoint &center )
...
+        for( gx = 0; gx < my_MAPSIZE * SEEX; ++gx ) {
+            for( gy = 0; gy < my_MAPSIZE * SEEY; ++gy ) {
+                total_grid += 1;
```
Comment (codemime) [inline src/map.cpp]: There's no need to increment here. This value can be calculated in compile time outside the loops as `my_MAPSIZE * SEEX * my_MAPSIZE * SEEY`

File: src/map.cpp [hunk 10]
```diff
@@ -5652,6 +5656,79 @@ void map::draw( const catacurses::window &w, const tripoint &center )
...
+                const auto grid_offset = point {gx, gy};
+                const tripoint map_location = tripoint( grid_offset, 0 );
+                if( ( !ter( map_location )->has_flag( TFLAG_INDOORS ) ) &&
+                    ( !ter( map_location )->has_flag( "TREE" ) ) && ( is_outside( map_location ) ) &&
+                    ( furn( map_location ) == f_null ) ) {
+                    if( one_in( std::max( 2, static_cast<int>( snowfactor / 2 ) ) ) ) {
```
Comment (codemime) [inline src/map.cpp]: `std::max( 2, static_cast<int>( snowfactor / 2 ) )` can be extracted from the loop as well.

File: src/map.cpp [hunk 11]
```diff
@@ -5652,6 +5656,79 @@ void map::draw( const catacurses::window &w, const tripoint &center )
...
+                        furn_set( map_location, f_snow );
+                        snowfall_settled += 1;
```
Comment (codemime) [inline src/map.cpp]: `++snowfall_settled` would be more conventional.

File: src/map.cpp [hunk 12]
```diff
@@ -5652,6 +5656,79 @@ void map::draw( const catacurses::window &w, const tripoint &center )
...
+                    }
+                }
+            }
+        }
+        percent_settled = std::min( 100, static_cast<int>( ( snowfall_settled / total_grid ) * 100 ) );
```
Comment (codemime) [inline src/map.cpp]: 1. `snowfall_settled` increments only when there was no snow on the tile initially and it was randomly placed there. Which means it denotes the amount of snow placed per invocation of `add_snow`, not the actual ratio. This seems to be wrong.

2. `snowfall_settled / total_grid` isn't what you would expect. Both are integers and produce either `0` or `1` (nothing else in the current implementation). Should be `static_cast<float>(snowfall_settled) / total_grid` to produce decimals.

3. `std::min( 100, ...)` is excessive, the ratio never exceeds `1`.

File: src/map.cpp [hunk 13]
```diff
@@ -5652,6 +5656,79 @@ void map::draw( const catacurses::window &w, const tripoint &center )
...
+                    if( furn( map_location ) == f_snow ) {
+                        if( one_in( tempfactor * std::max( 1, static_cast<int>( snowfall / 4000 ) ) ) ) {
+                            furn_set( map_location, f_null );
+                            snowfall_settled -= 1;
+                        }
+                    }
+                }
+            }
+        }
+        percent_settled = std::max( 0, static_cast<int>( ( snowfall_settled / total_grid ) * 100 ) );
```
Comment (codemime) [inline src/map.cpp]: Same here. Also, these functions look quite alike. Can their guts be deduplicated?

File: src/iexamine.cpp [hunk 14]
```diff
@@ -3044,6 +3045,22 @@ void iexamine::water_source(player &p, const tripoint &examp)
...
+{
+    if(!query_yn(_("Gather snow?"))) {
+        none( p, examp );
+        return;
+    }
+    if( p.can_pickWeight( item( "snow" ), true ) && p.can_pickVolume( item( "snow" ), true ) ){
+        p.i_add( item( "snow" ) );
+        p.add_msg_if_player( _( "You gathered some snow" ) );
+    } else {
+        g->m.add_item_or_charges( p.pos(), item( "snow" ) );
```
Comment (codemime) [inline src/iexamine.cpp]: `item` gets constructed four times (isn't quite cheap) from the same string literal. I'd extract a temporary for both deduplication and performance.

File: src/iexamine.cpp [hunk 15]
```diff
@@ -3044,6 +3045,22 @@ void iexamine::water_source(player &p, const tripoint &examp)
     g->handle_liquid( water, nullptr, 0, &examp );
 }
 
+void iexamine::gather_snow(player &p, const tripoint &examp)
+{
+    if(!query_yn(_("Gather snow?"))) {
```
Comment (codemime) [inline src/iexamine.cpp]: Coding style.

File: src/map.cpp [hunk 16]
```diff
@@ -6506,6 +6583,19 @@ void map::loadn( const int gridx, const int gridy, const int gridz, const bool u
...
     }
+    // Adding snow to new submaps
+    for( int j = 0; j < SEEY; j++ ) {
+        for( int i = 0; i < SEEX; i++ ) {
+            const tripoint pnt( gridx * SEEX + i, gridy * SEEY + j, gridz );
+            const point p( i, j );
+            if( ( !tmpsub->get_ter( p ).obj().has_flag( TFLAG_INDOORS ) ) &&
+                ( !tmpsub->get_ter( p ).obj().has_flag( "TREE" ) )  &&
+                ( tmpsub->get_furn( p ) == f_null ) && ( is_outside( pnt ) ) &&
+                x_in_y( percent_settled, 100 ) ) {
```
Comment (codemime) [inline src/map.cpp]: 1. Excessive parenthesis.
2. `has_flag` are presumably the most expensive checks here, they should go last.
3. `obj()` can be omitted, simply `get_ter( p )->has_flag( TFLAG_INDOORS )` would work too.
4. Not quite sure about the `"TREE"` flag (looks like a hack which doesn't cover all possible cases).

File: src/map.cpp [hunk 17]
```diff
@@ -6506,6 +6583,19 @@ void map::loadn( const int gridx, const int gridy, const int gridz, const bool u
...
+                tmpsub->frn[i][j] = furn_id( "f_snow" );
```
Comment (codemime) [inline src/map.cpp]: Please use `set_furn()` instead. It has an important side-effect (`is_uniform = false;`). Low-level code is best avoided here.

File: src/player.h [hunk 18]
```diff
@@ -1744,6 +1744,7 @@ class player : public Character
 
         /** Can the player lie down and cover self with blankets etc. **/
         bool can_use_floor_warmth() const;
+        bool walking_over_snow() const;
```
Comment (codemime) [inline src/player.h]: This function bloats the `player` class. I'd suggest getting rid of it.

File: src/weather.cpp [hunk 19]
```diff
@@ -166,16 +176,18 @@ weather_sum sum_conditions( const time_point &start, const time_point &end,
...
         if( g->weather_override == WEATHER_NULL ) {
             wtype = wgen.get_weather_conditions( location, t, g->get_seed() );
         } else {
             wtype = g->weather_override;
         }
         proc_weather_sum( wtype, data, t, tick_size );
+        count += 1;
     }
-
+    data.averagetemperature = static_cast<int>( data.averagetemperature / count );
```
Comment (codemime) [inline src/weather.cpp]: This code is also suspicious: `data.averagetemperature / count` follows the integer division rules and can't possibly have a fractional part, casting to `int` is excessive, it's already an `int`. You might want `static_cast<float>(data.averagetemperature) / count` or `averagetemperature` being `float` initially.

File: src/player.cpp [hunk 20]
```diff
@@ -1042,6 +1042,7 @@ void player::update_bodytemp()
     const bool has_climate_control = in_climate_control();
     const bool use_floor_warmth = can_use_floor_warmth();
     const furn_id furn_at_pos = g->m.furn( pos() );
+    const bool snowonground = walking_over_snow();
```
Comment (codemime) [inline src/player.cpp]: This temp is declared way too far from the place of utilization.

File: src/player.cpp [hunk 21]
```diff
@@ -1159,7 +1160,53 @@ void player::update_bodytemp()
         }
         // Loss of blood results in loss of body heat, 1% bodyheat lost per 2% hp lost
         temp_conv[bp] -= blood_loss( bp ) * temp_conv[bp] / 200;
-
+        bool waterproof_shoes = false;
+        for( const item &worn_item : worn ) {
+            if (worn_item.covers(bp) && worn_item.has_flag("WATERPROOF") ){
+                waterproof_shoes = true;
+            }
+        }
```
Comment (codemime) [inline src/player.cpp]: There's the `worn_with_flag` function for this.

File: src/player.cpp [hunk 22]
```diff
@@ -1159,7 +1160,53 @@ void player::update_bodytemp()
...
+                if( snowonground ){
+                    if( waterproof_shoes ){
+                        break;
+                    } else if( wearing_shoes || mutfeet ){
+                        temp_conv[bp] -= 300;
+                    } else {
+                        temp_conv[bp] -= 450;
+                    }
+                }
+                break;
```
Comment (codemime) [inline src/player.cpp]: Duplicated cases:
```c++
case bp_foot_l:
case bp_foot_r:
    ...
```
You probably don't want a gigantic `switch` statement here since you're only interested in feet. A couple of conditions in an `if` would suffice.

**Comment by davidpwbrown** [comment]
Yeah, i'm a fix up the horrible code at some point, I'm trying to think of ways to solve the difficult problems of the visuals/tiles. then I'll fix it up the rest..

**Comment by Throwaway-name** [comment]
Would this work with this?: https://github.com/CleverRaven/Cataclysm-DDA/pull/27387

**Comment by davidpwbrown** [comment]
> Would this work with this?: #27387

Possibly. eventually I want to track water on the ground from rainwater, and the thermal energy system could potentially be used to freeze  puddles/rivers and let snow settle. I'll come back to this and similar things when I get better ideas for how to implement it.

**Comment by davidpwbrown** [comment]
a project for another time. perhaps.

**Comment by kevingranade** [comment]
This pull request has been mentioned on **Cataclysm: Dark Days Ahead**. There might be relevant details there:

https://discourse.cataclysmdda.org/t/winter-activities-melting-snow-for-water-snow-taffy-building-snowmen-throwing-snowballs/20579/6

**Comment by ThomasLinkin** [comment]
so what happened to this concept now since DeadPeople tileset died months ago ?

**Comment by anothersimulacrum** [comment]
What? Tilesets have very very little to do with this concept.
